src/Workspaces/Core/Portable/Shared/Utilities/AsyncBatchingWorkQueue.cs

```diff
@@ -43,37 +46,43 @@ internal class AsyncBatchingWorkQueue<TItem>
         #region protected by lock
 
         /// <summary>
-        /// Lock we will use to ensure the remainder of these fields can be accessed in a threadsafe
-        /// manner.  When work is added we'll place the data into <see cref="_nextBatch"/>.
-        /// We'll then kick of a task to process this in the future if we don't already have an
-        /// existing task in flight for that.
+        /// Lock we will use to ensure the remainder of these fields can be accessed in a threadsafe manner.  When work
+        /// is added we'll place the data into <see cref="_nextBatchArray"/> or <see cref="_nextBatchMap"/> We'll then
+        /// kick of a task to process this in the future if we don't already have an existing task in flight for that.
         /// </summary>
         private readonly object _gate = new object();
 
         /// <summary>
         /// Data added that we want to process in our next update task.
         /// </summary>
-        private readonly ArrayBuilder<TItem> _nextBatch = ArrayBuilder<TItem>.GetInstance();
+        private readonly ArrayBuilder<TItem> _nextBatchArray = ArrayBuilder<TItem>.GetInstance();
 
         /// <summary>
-        /// Used if <see cref="_equalityComparer"/> is present to ensure only unique items are added to <see
-        /// cref="_nextBatch"/>.
+        /// Used if <see cref="_equalityComparer"/> is present to ensure only unique items are processed. The value in
+        /// the dictionary is the index of the item so we can process them in the order we heard about them.
         /// </summary>
-        private readonly HashSet<TItem> _uniqueItems;
+        private readonly Dictionary<TItem, int>? _nextBatchMap;
```

Comment:
the previous system allowed deduping items in the queue, however it had an impl where the *first* item added ot hte queue was the one that would be processed.  In practice this was fine (and not observable) because the features adding items were adding identical value types.  So whether you processed the item originally in the list, or hte new one being added wouldn't make a difference.

I wanted the semantics here of processing a batch of Documents.  However, given a Document for Foo.cs if i heard about multiple changes to it, i wanted to keep adding that same Document, just with the wrinkle that we would process the *last* item added, not hte first.  This necessitated this change in the async work queue.  Previously we just had a set, and we would only add to the array if we were able to add to the set.  Now that we want the last one to win, we have to keep a dictionary so we can map from the item to it's position in the array so we can replace it with the new item.